File: kimi-k3/reference/python/kimi_k3_ref/rmsnorm.py
"""RMSNorm matching ``KimiRMSNorm`` in modeling_kimi_linear.py."""

from __future__ import annotations

import numpy as np


def rmsnorm(x: np.ndarray, weight: np.ndarray, eps: float = 1e-5) -> np.ndarray:
    """
    Parameters
    ----------
    x : float32 array, shape (..., hidden)
    weight : float32 array, shape (hidden,)
    eps : variance epsilon (official default rms_norm_eps=1e-5)

    Official:
        x = x.float()
        x = x * rsqrt(x.pow(2).mean(-1, keepdim=True) + eps)
        return weight * x.to(dtype)
    """
    x = np.asarray(x, dtype=np.float32)
    weight = np.asarray(weight, dtype=np.float32)
    var = np.mean(np.square(x), axis=-1, keepdims=True)
    y = x * np.reciprocal(np.sqrt(var + np.float32(eps)))
    return (weight * y).astype(np.float32)

File: kimi-k3/rust/src/rmsnorm.rs
/// RMSNorm matching `KimiRMSNorm`.
pub fn rmsnorm(x: &[f32], weight: &[f32], eps: f32, out: &mut [f32]) {
    assert_eq!(x.len(), out.len());
    let hidden = weight.len();
    assert!(hidden > 0 && x.len() % hidden == 0);
    let rows = x.len() / hidden;
    for r in 0..rows {
        let base = r * hidden;
        let mut var = 0.0f32;
        for j in 0..hidden {
            let v = x[base + j];
            var += v * v;
        }
        var /= hidden as f32;
        let inv = (var + eps).sqrt().recip();
        for j in 0..hidden {
            out[base + j] = x[base + j] * inv * weight[j];
        }
    }
}
